python/cudf_polars/cudf_polars/experimental/io.py

```diff
@@ -421,3 +412,263 @@ def _(
+        ):
+            row_group_unique_count = plc.stream_compaction.distinct_count(
+                column,
+                plc.types.NullPolicy.INCLUDE,
+                plc.types.NanPolicy.NAN_IS_NULL,
+            )
+            result = {
+                "exact": exact,
+                "fraction": row_group_unique_count / row_group_num_rows,
+            }
+            # Assume that if every row is unique then this is a
+            # primary key otherwise it's a foreign key and we
+            # can't use the single row group count estimate
+            # Example, consider a "foreign" key that has 100
+            # unique values. If we sample from a single row group,
+            # we likely obtain a unique count of 100. But we can't
+            # necessarily deduce that that means that the unique
+            # count is 100 / num_rows_in_group * num_rows_in_file
+            if exact:
+                result["count"] = row_group_unique_count
+            elif row_group_unique_count == row_group_num_rows:
+                result["count"] = self._row_count.value
+
+            self._unique_stats[name] = UniqueInfo(**result)
+
+    @property
+    def row_count(self) -> RowCountInfo:
+        """Data source row-count estimate."""
+        if self.max_file_samples < 1:
+            return RowCountInfo()  # pragma: no cover
+
+        if self._row_count is None:
+            self._sample_metadata()
+
+        assert self._row_count is not None, "metadata sampling failed."
+        return self._row_count
+
+    def unique(self, column: str) -> UniqueInfo:
+        """Return unique-value information."""
+        if self.max_file_samples < 1 or self.max_rg_samples < 1:
+            return UniqueInfo()  # pragma: no cover
+
+        if column not in self._unique_stats:
+            self.add_unique_stats_column(column)
+            self._sample_row_groups()
+            self._key_columns = set()
+
+        return self._unique_stats.get(column, UniqueInfo())
+
+    def storage_size(self, column: str) -> StorageSizeInfo:
+        """Return the average column size for a single file."""
+        if self.max_file_samples < 1:
+            return StorageSizeInfo()  # pragma: no cover
+
+        if self._mean_size_per_file is None:
+            self._sample_metadata()
+
+        assert self._mean_size_per_file is not None, "metadata sampling failed"
+        return self._mean_size_per_file.get(column, StorageSizeInfo())
+
+    def add_unique_stats_column(self, column: str) -> None:
+        """Add a column needing unique-value information."""
+        if column not in self._key_columns and column not in self._unique_stats:
+            self._key_columns.add(column)
+
+
+@functools.lru_cache(maxsize=10)
+def _sample_pq_stats(
+    paths: tuple[str, ...],
+    max_file_samples: int,
+    max_rg_samples: int,
+) -> PqSourceInfo:
+    """Return Parquet datasource information."""
+    return PqSourceInfo(paths, max_file_samples, max_rg_samples)
```

Comment:
I kind of think that this cache should be per query.

Rationale, between queries I don't think we can guarantee that the statistics are valid, because the files might be overwritten.

WDYT?

Reply:
Sorry I didn't comment on this explicitly yet.

You are correct that we need to cache a bit more than a tuple of paths if we want to account for over-written data. Dask tries to do this [with a checksum](https://github.com/dask/dask/blob/a7e05043816c6243d9b69e6bec09d61cb078b957/dask/dataframe/dask_expr/io/parquet.py#L1394). We can start by clearing the cache for every new query, but I'd love to find a way to preserve the cache between queries (**very** useful for remote data in practice).

Reply:
Yeah, the current implementation will clear the caches every time we call `evaluate_streaming`. Hopefully that is  reasonable for now?